### PR TITLE
Path functions and secondary files / directory improvements

### PR DESCRIPTION
**Pull request by Horneth**
- Cleans up the `IoFunctionSet` a little and add `PathFunctionSet` in it to deal with path manipulation that doesn't involve I/O
- When secondary files are actually secondary directories, list their content and return a `WomMaybeListedDirectory` instead of a `WomSingleFile` or `WomMaybePopulatedFile`
- Make paths absolute as much and as early as possible to avoid ambiguity and having to guess later where they are relative to
- Adjust the `OutputManipulator` to deal with secondary listed directories

### CODE REVIEW

**Comment by mcovarr** [comment]
Could you please give a very brief overview of the conceptual changes here?

File: centaurCwlRunner/src/main/scala/centaur/cwl/OutputManipulator.scala [hunk 1]
```diff
@@ -85,39 +101,39 @@ object OutputManipulator extends Poly1 {
         * 2) the size calculated from the value of the "contents" field of the metadata
         * 3) the size calculated from the file itself
         */
-      val defaultSize = sizeContent.orElse(sizeFile).map(Json.fromLong).getOrElse(Json.Null)
+      val defaultSize = sizeContent.orElse(sizeFile(path)).map(Json.fromLong).getOrElse(Json.Null)
       val size = valueOrNull("size", defaultSize)
 
+      val basename: Option[Json] = if (isInsideDirectory) {
+        Option(valueOrNull("basename", path.exists.option(path.nameWithoutExtension).map(Json.fromString).getOrElse(Json.Null)))
+      } else None
```
Comment (mcovarr) [inline centaurCwlRunner/src/main/scala/centaur/cwl/OutputManipulator.scala]: I'm not getting the "inside directory" logic here

File: cwl/src/main/scala/cwl/CwlType.scala [hunk 2]
```diff
@@ -107,18 +110,44 @@ object File {
...
 
   def secondaryExpressionFiles(primaryWomFile: WomFile,
                                stringWomFileType: WomFileType,
                                expression: Expression,
                                parameterContext: ParameterContext,
-                               expressionLib: ExpressionLib): ErrorOr[List[WomFile]] = {
+                               expressionLib: ExpressionLib,
+                               ioFunctions: IoFunctionSet): ErrorOr[List[WomFile]] = {
+    
+    def relativize(path: String) = ioFunctions.pathFunctions.absoluteSibling(primaryWomFile.value, path)
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/CwlType.scala]: `relativize` returns the result of `absoluteSibling`?

File: src/bin/travis/resources/local_conformance_expected_failures.txt [hunk 3]
```diff
@@ -88,10 +88,10 @@
...
 #90
-91
+#91
 92
 #93
-94
+#94
 #95
 96
 #97
```
Comment (mcovarr) [inline src/bin/travis/resources/local_conformance_expected_failures.txt]: For rebase these are 55, 91 and 99

File: src/bin/travis/resources/papi_conformance_expected_failures.txt [hunk 4]
```diff
@@ -79,7 +79,7 @@
 79
 80
 81
-82
+#82
 83
 84
 #85
```
Comment (mcovarr) [inline src/bin/travis/resources/papi_conformance_expected_failures.txt]: 56 and 87

File: backend/src/main/scala/cromwell/backend/wdl/WriteFunctions.scala [hunk 5]
```diff
@@ -33,7 +30,7 @@ trait WriteFunctions extends PathFactory with IoFunctionSet with AsyncIoFunction
 
   private val relativeToLocal = System.getProperty("user.dir").ensureSlashed
 
-  def relativeToAbsolutePath(pathFrom: String): String = if (new File(pathFrom).isAbsolute) pathFrom else relativeToLocal + pathFrom
+  def relativeToAbsolutePath(pathFrom: String): String = if (buildPath(pathFrom).isAbsolute) pathFrom else relativeToLocal + pathFrom
```
[the remark this answers is not in the document]
Comment (Horneth) [reply backend/src/main/scala/cromwell/backend/wdl/WriteFunctions.scala]: yes, and that's why we ended up with weird concatenated gcs paths

File: centaurCwlRunner/src/main/scala/centaur/cwl/OutputManipulator.scala (hunk 1 above)
Comment (Horneth) [reply centaurCwlRunner/src/main/scala/centaur/cwl/OutputManipulator.scala]: Yeah I meant to add a comment and forgot. Apparently the conformance tests expect the files listed in the `listing` field for directories to have `basename` filled in, but only there and not for other top level files...
So this flag is there to decide if we should add basename

File: cwl/src/main/scala/cwl/CwlType.scala (hunk 2 above)
Comment (Horneth) [reply cwl/src/main/scala/cwl/CwlType.scala]: Yeah the naming is confusing -_-
What it's doing is taking a "relative" path A and turning it into an absolute path B. It's doing that by saying, A is actually relative to absolute path C, so the absolute version of A is C/A.
Not sure how to describe that in terms of function names 😄

File: backend/src/main/scala/cromwell/backend/io/DirectoryFunctions.scala [hunk 6]
```diff
@@ -41,25 +46,7 @@ trait DirectoryFunctions extends IoFunctionSet with PathFactory {
...
-      } else {
-        newCheckedPaths.valid
-      }
-    }
-
-    val allPaths = listPaths(path, Set.empty).map(_.toList)
-    val allFiles = allPaths.map(_.filterNot(_.isDirectory))
-    allFiles
-  }
+  def listFiles(path: Path): ErrorOr[List[Path]] = path.listRecursively.filterNot(_.isDirectory).toList.validNel
```
Comment (danbills) [inline backend/src/main/scala/cromwell/backend/io/DirectoryFunctions.scala]: :100:

**Pull request merged**